app/controllers/conversations_controller.rb

```diff
@@ -2,7 +2,7 @@ class ConversationsController < ApplicationController
   before_action :authenticate_user!
 
   def index
-    @conversations = current_user.conversations.order(updated_at: :desc)
+    @conversations = current_user.conversations.includes(:business).order(updated_at: :desc)
```

Comment:
Do we still need this if we are using the IDs?

Reply:
Unfortunately we do. Otherwise the `business?` method in the  `Conversation` model would shoot off n+1 queries. This should effectively make it so that computing every conversations `read_status` will take 2 queries. Unless I'm off with that 🤔

Reply:
Hmm. Since the notifications are scoped to a user (not a business or developer) what if we track the read status based on the user?

Reply:
That was my first thought as well. Honestly have no idea why I didn't go with that approach first time around, lol. I think it was 4am when I took a first stab? lol.

Something akin to
```
current_user.conversations.order(updated_at: :desc).select(:'conversations.*', 
'EXISTS(SELECT 1 FROM notifications WHERE params @> conversations AND conversations.recipient_type = 'User' AND conversations.recipient_id = #{current_user.id} AND read_at IS NULL) AS unread_messages`
```

Syntax is wrong but that's the general idea. I think 4am me forgot about extending select? Or is this an antipattern, lol.

Either way, I would translate this to an active query type command. Just wanted to get your thoughts before re-digging in.

Reply:
I'm finding that SQL statement pretty hard to read, but if we can convert it to Active Record or Arel then that sounds like a better approach to me!

Reply:
Ah, a bit SOL. I remember the problem now. We needs rails to serialize the object to query by it, which isn't supported OOTB by noticed, because it really isn't supported OOTB by rails as far as I can tell (happy to be proved wrong!). That's bc even though it's scoped to the user, the notifications params is where we need to find the relevant conversation to know whether or not to check the `read_at` column. 

Trying something, fingers crossed! If I can make it work I'll be busy submitting a PR for noticed gem :)

Reply:
The reason for this is because the params are serialized with the `ActiveJob` serializer. Great for the DB, was never really built to be query-able.

Reply:
Have a solution up, take a look and let me know what you have qualms with. Sort of had to force it to make it work with two queries, but it's now scoped to the user.

Reply:
Sorry, but I feel like I'm missing something...

If we are caching the unread status via `conversations.read_status` then why do we need a complex query? You've already added the callback when sending a message to update the conversation.

Why can't we iterate over conversations and then:

```erb
<% conversations.each do |conversation| %>
  <h2>Conversation with <%= conversation.other_recipient(current_user).name %></h2>
  <% if conversation.business?(current_user) && conversation. messages_for_business_unread? || conversation.developer?(current_user) && conversation.messages_for_developer_unread? %>
    <p>Unread messages</p>
  <% end %
<% end %>
```

Reply:
What a d'oh moment! Way more straightforward.